src/Firestore/Query.php

```diff
@@ -560,47 +598,166 @@ private function buildPosition(array $fieldValues, $before)
 
-                if ($value->parent()->name() !== $collection) {
-                    throw new \BadMethodCallException(
+                $parent = ($value instanceof DocumentReference)
+                    ? $value->parent()->name()
+                    : $value->reference()->parent()->name();
+
+                if ($parent !== $collection) {
+                    throw new \InvalidArgumentException(
                         'Only direct children may be used as query boundaries.'
                     );
                 }
             }
         }
 
+        return $this->newQuery([
+            $key => [
+                'before' => $before,
+                'values' => $this->valueMapper->encodeValues($fieldValues)
+            ],
+            'orderBy' => $orderBy
+        ], true);
+    }
+
+    /**
+     * Build cursors for document snapshots.
+     *
+     * @param DocumentSnapshot $snapshot The document snapshot
+     * @return array[] A list, where position 0 is fieldValues and position 1 is orderBy.
+     */
+    private function snapshotPosition(DocumentSnapshot $snapshot, array $orderBy)
+    {
+        $append = true;
+        $direction = self::DIR_ASCENDING;
+
+        $fieldValues = [];
+
+        // If the list of orderBy clauses already contains __name__, use it unchanged.
+        $append = !(bool) array_filter($orderBy, function ($order) {
+            return $order['field']['fieldPath'] === self::DOCUMENT_ID;
+        });
+
+        // If there is inequality filter (anything other than equals),
+        // append orderBy(the last inequality filter’s path, ascending).
+        if (!$orderBy && $this->queryHas('where') && $append) {
+            $filters = $this->query['where']['compositeFilter']['filters'];
+            $inequality = array_filter($filters, function ($filter) {
+                $type = array_keys($filter)[0];
+                return $filter[$type]['op'] !== self::OP_EQUAL;
+            });
+
+            if ($inequality) {
+                $filter = end($inequality);
+                $type = array_keys($filter)[0];
+                $orderBy[] = [
+                    'field' => [
+                        'fieldPath' => $filter[$type]['field']['fieldPath'],
+                    ],
+                    'direction' => self::DIR_ASCENDING
+                ];
+            }
+        }
+
+        // If the query has existing orderBy constraints
+        if ($orderBy) {
+            // If the list of orderBy clauses already contains __name__, use it unchanged.
+            $append = !(bool) array_filter($orderBy, function ($order) {
+                return $order['field']['fieldPath'] === self::DOCUMENT_ID;
+            });
+
+            // Append orderBy(__name__, direction of last orderBy clause)
+            if ($append) {
+                $lastOrderDirection = end($orderBy)['direction'];
+                $fieldValues[] = $snapshot;
+                $orderBy[] = [
+                    'field' => [
```

Comment:
Is this appending?

Furthermore, you need to return a new copy since Query instances are immutable.

Reply:
> Furthermore, you need to return a new copy since Query instances are immutable.

`$orderBy` is passed into this method by value, and does not mutate the parent state. It is returned in a new Query instance out of the calling method, `buildPosition`.